src/mono/mono/mini/interp/transform-opt.c

```diff
@@ -481,45 +484,1072 @@ interp_alloc_offsets (TransformData *td)
+				out_bb->dfs_index = -2;
+			}
+		}
+	}
+
+	g_free (stack);
+	return dfs_index;
+}
+
+static void
+interp_compute_dfs_indexes (TransformData *td)
+{
+	for (InterpBasicBlock *bb = td->entry_bb; bb != NULL; bb = bb->next_bb)
+		bb->dfs_index = -1;
+	// Sort bblocks in reverse postorder
+	int dfs_index = dfs_visit (td);
+	td->bblocks_count = dfs_index;
+
+	// Visit also bblocks reachable from eh handlers. These bblocks are not linked
+	// to the main cfg (where we do dominator computation, ssa transformation etc)
+	if (td->header->num_clauses > 0) {
+		InterpBasicBlock *current = td->entry_bb;
+		while (current != NULL) {
+			if (current->reachable && current->dfs_index == -1) {
+				current->dfs_index = dfs_index;
+				td->bblocks [dfs_index] = current;
+				dfs_index++;
+			}
+			current = current->next_bb;
+		}
+	}
+	td->bblocks_count_eh = dfs_index;
+
+	if (td->verbose_level) {
+		InterpBasicBlock *bb;
+		g_print ("\nBASIC BLOCK GRAPH:\n");
+		for (bb = td->entry_bb; bb != NULL; bb = bb->next_bb) {
+			GString* bb_info = interp_get_bb_links (bb);
+			g_print ("BB%d: DFS%s(%d), %s\n", bb->index, (bb->dfs_index >= td->bblocks_count) ? "_EH" : "" , bb->dfs_index, bb_info->str);
+			g_string_free (bb_info, TRUE);
+		}
+	}
+}
+
+static InterpBasicBlock*
+dom_intersect (InterpBasicBlock **idoms, InterpBasicBlock *bb1, InterpBasicBlock *bb2)
+{
+	while (bb1 != bb2) {
+		while (bb1->dfs_index < bb2->dfs_index)
+			bb2 = idoms [bb2->dfs_index];
+		while (bb2->dfs_index < bb1->dfs_index)
+			bb1 = idoms [bb1->dfs_index];
+	}
+	return bb1;
+}
+
+static gboolean
+is_bblock_ssa_cfg (TransformData *td, InterpBasicBlock *bb)
+{
+	// bblocks with uninitialized dfs_index are unreachable
+	if (bb->dfs_index == -1)
+		return FALSE;
+	if (bb->dfs_index < td->bblocks_count)
+		return TRUE;
+	return FALSE;
+}
+
+static void
+interp_compute_dominators (TransformData *td)
+{
+	InterpBasicBlock **idoms = (InterpBasicBlock**)mono_mempool_alloc0 (td->opt_mempool, sizeof (InterpBasicBlock*) * td->bblocks_count);
+
+	idoms [0] = td->entry_bb;
+	gboolean changed = TRUE;
+	while (changed) {
+		changed = FALSE;
+		// all bblocks in reverse post order except entry
+		for (int i = 1; i < td->bblocks_count; i++) {
+			InterpBasicBlock *bb = td->bblocks [i];
+			InterpBasicBlock *new_idom = NULL;
+			// pick candidate idom from first processed predecessor of it
+			int j;
+			for (j = 0; j < bb->in_count; j++) {
+                                InterpBasicBlock *in_bb = bb->in_bb [j];
+                                if (is_bblock_ssa_cfg (td, in_bb) && idoms [in_bb->dfs_index]) {
+                                        new_idom = in_bb;
+                                        break;
+                                }
+                        }
+
+			// intersect new_idom with dominators from the other predecessors
+			for (; j < bb->in_count; j++) {
+				InterpBasicBlock *in_bb = bb->in_bb [j];
+				if (is_bblock_ssa_cfg (td, in_bb) && idoms [in_bb->dfs_index])
```

Comment:
is it possible for `in_bb->dfs_index` to be -1 here? I got the impression some BBs could be unreachable and it would be -1, but it's not obvious to me whether we are certain by construction that we will never have one of those in `bb->in_bb[]`